openmls/src/group/mls_group/new_from_external_init.rs

```diff
@@ -0,0 +1,174 @@
+use mls_group::create_commit_params::CommitType;
+
+use crate::{
+    ciphersuite::signable::Verifiable,
+    credentials::CredentialBundle,
+    group::errors::ExternalInitError,
+    messages::{
+        proposals::{ExternalInitProposal, Proposal},
+        public_group_state::{PublicGroupState, VerifiablePublicGroupState},
+    },
+    prelude::{plaintext::MlsPlaintext, KeyPackageBundle},
+};
+
+use super::{
+    create_commit_params::CreateCommitParams,
+    proposals::{ProposalStore, StagedProposal},
+    CoreGroup,
+};
+use crate::group::mls_group::*;
+
+pub type ExternalInitResult = Result<
+    (
+        CoreGroup,
+        MlsPlaintext,
+        Option<Welcome>,
+        Option<KeyPackageBundle>,
+    ),
+    ExternalInitError,
+>;
+
+impl CoreGroup {
+    /// Join a group without the help of an internal member. This function
+    /// requires a `PublicGroupState`, as well as the corresponding public tree
+    /// `nodes`. After the group state is initialized, this function creates an
+    /// `ExternalInit` proposal and commits it along with the given proposals by
+    /// reference and by value.
+    ///
+    /// Returns the new `CoreGroup` object, as well as the `MlsPlaintext`
+    /// containing the commit.
+    pub fn new_from_external_init(
+        backend: &impl OpenMlsCryptoProvider,
+        framing_parameters: FramingParameters,
+        nodes_option: Option<&[Option<Node>]>,
+        credential_bundle: &CredentialBundle,
+        proposals_by_reference: &[MlsPlaintext],
+        proposals_by_value: &[Proposal],
+        verifiable_public_group_state: VerifiablePublicGroupState,
+    ) -> ExternalInitResult {
+        let ciphersuite = Config::ciphersuite(verifiable_public_group_state.ciphersuite())?;
+        if !Config::supported_versions().contains(&verifiable_public_group_state.version()) {
+            return Err(ExternalInitError::UnsupportedMlsVersion);
+        }
+
+        let mut ratchet_tree_extensions = verifiable_public_group_state
+            .other_extensions()
+            .iter()
+            .filter(|e| e.extension_type() == ExtensionType::RatchetTree)
+            .collect::<Vec<&Extension>>();
+
+        let ratchet_tree_extension = if ratchet_tree_extensions.is_empty() {
+            None
+        } else if ratchet_tree_extensions.len() == 1 {
```

Comment:
Just pop out the first element then you don' need to check the length

Reply:
I changed my mind and we are now detecting duplicate extensions again. However, that requires that I use `filter` instead of find here to make sure we're catching all of them (because you mentioned it in another comment..